### PR TITLE
wiseconnect: Add Bluetooth HCI driver

### PR DESCRIPTION
**Pull request by silabs-TiborL**

### CODE REVIEW

File: wiseconnect/components/device/silabs/si91x/mcu/drivers/unified_api/src/sl_si91x_driver_gpio.c
Comment (asmellby) [inline wiseconnect/components/device/silabs/si91x/mcu/drivers/unified_api/src/sl_si91x_driver_gpio.c]: This is fixed in a different way in Wiseconnect 3.3.1. It would be cleaner if we apply the exact same change as upstream for now:

```
  if (!(gpio_num >= HOST_PAD_MIN && gpio_num <= HOST_PAD_MAX)) {
```

Also, patches should preferably be done in separate commits from the ones introducing unmodified code from upstream.

We also want to upgrade to 3.3.2 soon, but let's keep that separate from this PR.

**Review by asmellby — COMMENTED** [review]
Please make sure that your commit message contains origin information including the git hash from which the content was imported. See https://docs.zephyrproject.org/latest/contribute/guidelines.html#identifying-contribution-origin for instructions, or any of the previous commits introducing HAL content.

File: wiseconnect/components/device/silabs/si91x/mcu/drivers/unified_api/src/sl_si91x_driver_gpio.c
Comment (silabs-TiborL) [reply wiseconnect/components/device/silabs/si91x/mcu/drivers/unified_api/src/sl_si91x_driver_gpio.c]: I did not want to add this to this PR, will revert.

**Comment by jhedberg** [comment]
@silabs-TiborL you need to rebase, since #7 introduced changes to `CMakeLists.txt` that conflicts with this one.

File: wiseconnect/components/device/silabs/si91x/wireless/inc/sl_wifi_device.h
Comment (jhedberg) [inline wiseconnect/components/device/silabs/si91x/wireless/inc/sl_wifi_device.h]: Is this purely a downstream change in this tree, or was the same fix also submitted to the upstream wiseconnect tree? If it was not submitted upstream, we'll need to keep making sure to always fix this as we move to newer wiseconnect versions. Btw, the code comment above this line still says "Extention" :)

Comment (silabs-TiborL) [reply wiseconnect/components/device/silabs/si91x/wireless/inc/sl_wifi_device.h]: It's a pure downstream change, we could raise a ticket to WiseConnect team to fix this.

This typo is only present in the validity bits (which is currently the most significant bit).
So another possible solution would be to create new defines in the Zephyr repo with "BIT(31)". However if the SDK changes this can cause a "silent error" whereas the typo fix will result in a build error.

What would be your preference?

Comment (jhedberg) [reply wiseconnect/components/device/silabs/si91x/wireless/inc/sl_wifi_device.h]: We want to automate the importing of Silabs SDKs into `hal_silabs`. @asmellby has already done the initial work for the SiSDK in his latest PR. I think the cleanest and most sustainable option would be to raise a ticket to get this fixed upstream, since otherwise we would need to have extra logic in our automation to do the fix there. The option of having a custom define in Zephyr is the least preferred one, since we want to stay in sync with the upstream SDK and its API.

Comment (silabs-TiborL) [reply wiseconnect/components/device/silabs/si91x/wireless/inc/sl_wifi_device.h]: Okay, I'll leave it as is now and create a ticket about this.

**Review by jhedberg — APPROVED** [review]

**Comment by jhedberg** [comment]
@asmellby @jerome-pouiller could we get a second review here? Thanks!

**Review by asmellby — APPROVED** [review]

File: wiseconnect/CMakeLists.txt
Comment (jerome-pouiller) [inline wiseconnect/CMakeLists.txt]: Do we want to name this option `CONFIG_SIWX917_NWP_NETWORK_STACK` or `CONFIG_WISECONNECT_NETWORK_STACK`?

Comment (jhedberg) [reply wiseconnect/CMakeLists.txt]: Currently these are defined in `soc/silabs/silabs_siwx917/siwg917/Kconfig` in the main tree. If we change the name to `WISECONNECT_*` I think a more appropriate place would be `modules/Kconfig.silabs` (also in the main tree).

Comment (jerome-pouiller) [reply wiseconnect/CMakeLists.txt]: Yes, we can consider this option is more a "mode of operation" of WiseConnect rather than just a support for a driver. I feel it make sense to hide it in `modules/Kconfig.silabs` (note this file is currently named `modules/Kconfig` in our downstream)

Comment (jhedberg) [reply wiseconnect/CMakeLists.txt]: > Yes, we can consider this option is more a "mode of operation" of WiseConnect rather than just a support for a driver. I feel it make sense to hide it in `modules/Kconfig.silabs`

Sure, makes sense. Unrelated to the current case, but I have a feeling we may want to move to using `kconfig-ext` in `module.yml` and then use `modules/hal_silabs/Kconfig` in the main tree. That seems to be the more "modern" approach.

> note this file is currently named modules/Kconfig in our downstream

Hmm? I can't find a modules directory at all in zephyr-silabs.

Comment (silabs-TiborL) [reply wiseconnect/CMakeLists.txt]: Is it ok to do just the rename here and do the relocation in another PR?

Comment (jhedberg) [reply wiseconnect/CMakeLists.txt]: > Is it ok to do just the rename here and do the relocation in another PR?

So you agree this can be seen as a more WiseConnect specific thing rather than a 917 specific thing? I'm not opposed to just doing the rename at first, though in both cases you will anyway need to update your zephyr-silabs PR, i.e. it's not really that much additional work to create a `modules/Kconfig.silabs` that will eventually get merged with the corresponding file in the main tree.

Btw, I realized that we have `config HAS_SILABS_WISECONNECT` in the zephyr-silabs root Kconfig, which should really also be in the module Kconfig file.

Comment (jerome-pouiller) [reply wiseconnect/CMakeLists.txt]: >> note this file is currently named modules/Kconfig in our downstream
>
> [...]
>
> Btw, I realized that we have config HAS_SILABS_WISECONNECT in the zephyr-silabs root  Kconfig, which should really also be in the module Kconfig file.

Yes, it was what I wanted to say.

Comment (silabs-TiborL) [reply wiseconnect/CMakeLists.txt]: Yes, completely agree. 
You're right, I need to update both PR so then I'll do these Kconfig modifications as well.

Comment (jhedberg) [reply wiseconnect/CMakeLists.txt]: @silabs-TiborL I think there was some misunderstanding. The place I was suggesting to put this was in the `zephyr-silabs` tree and not the `hal_silabs` tree. The goal is also to remove all CMake and Kconfig files completely from `hal_silabs` and have those instead in the main tree (or `zephyr-silabs` in the case of pre-upstream 917 work).

For now I'd create a `modules/Kconfig.silabs` file in the `zephyr-silabs` tree and put all of this there. Please also see my comment in the #zephyr-development Slack channel how this should change in the long-term (however for now `modules/Kconfig.silabs` is the way to go, since that mirrors what's in the upstream main tree).

Comment (silabs-TiborL) [reply wiseconnect/CMakeLists.txt]: I see, sorry about this, I've only started working with Zephyr recently andI'm not 100% sure that I got your intention correctly. Thus I'd go for the easy path and just rename CONFIG_SIWX917_NWP_NETWORK_STACK to CONFIG_WISECONNECT_NETWORK_STACK for this PR.

**Review by jhedberg — APPROVED** [review]

File: wiseconnect/CMakeLists.txt
Comment (jhedberg) [reply wiseconnect/CMakeLists.txt]: Sure. Let's go with that then in the short-term.

**Comment by jhedberg** [comment]
@jerome-pouiller are you ok with the intermediate solution? As a follow-up PR we can reorganize the Kconfig options and files in zephyr-silabs.

**Review by jerome-pouiller — APPROVED** [review]

**Pull request merged**
